src/analysis/processing/qgsalgorithmrasterrank.cpp

```diff
@@ -0,0 +1,309 @@
+  }
+  else
+  {
+    outputCrs = mLayers.at( 0 )->crs();
+  }
+
+
+  const Qgis::DataType outputDataType = qobject_cast<QgsRasterLayer *>( mLayers.at( 0 ) )->dataProvider()->dataType( 1 );
+  double outputNoData = 0.0;
+  if ( qobject_cast<QgsRasterLayer *>( mLayers.at( 0 ) )->dataProvider()->sourceHasNoDataValue( 1 ) )
+  {
+    outputNoData = qobject_cast<QgsRasterLayer *>( mLayers.at( 0 ) )->dataProvider()->sourceNoDataValue( 1 );
+  }
+  else
+  {
+    outputNoData = -FLT_MAX;
+  }
+  const bool outputNoDataOverride = parameterAsInt( parameters, QStringLiteral( "NODATA_HANDLING" ), context ) == 1;
+
+  QgsRectangle outputExtent;
+  if ( parameters.value( QStringLiteral( "EXTENT" ) ).isValid() )
+  {
+    outputExtent = parameterAsExtent( parameters, QStringLiteral( "EXTENT" ), context, outputCrs );
+  }
+  else
+  {
+    outputExtent = QgsProcessingUtils::combineLayerExtents( mLayers, outputCrs, context );
+  }
+
+  double minCellSizeX = 1e9;
+  double minCellSizeY = 1e9;
+  for ( QgsMapLayer *layer : mLayers )
+  {
+    QgsRasterLayer *rLayer = qobject_cast<QgsRasterLayer *>( layer );
+
+    QgsRectangle extent = rLayer->extent();
+    if ( rLayer->crs() != outputCrs )
+    {
+      QgsCoordinateTransform ct( rLayer->crs(), outputCrs, context.transformContext() );
+      extent = ct.transformBoundingBox( extent );
+    }
+
+    minCellSizeX = std::min( minCellSizeX, ( extent.xMaximum() - extent.xMinimum() ) / rLayer->width() );
+    minCellSizeY = std::min( minCellSizeY, ( extent.yMaximum() - extent.yMinimum() ) / rLayer->height() );
+  }
+
+  double outputCellSizeX = parameterAsDouble( parameters, QStringLiteral( "CELL_SIZE" ), context );
+  double outputCellSizeY = outputCellSizeX;
+  if ( outputCellSizeX == 0 )
+  {
+    outputCellSizeX = minCellSizeX;
+    outputCellSizeY = minCellSizeY;
+  }
+
+  double cols = std::round( ( outputExtent.xMaximum() - outputExtent.xMinimum() ) / outputCellSizeX );
+  double rows = std::round( ( outputExtent.yMaximum() - outputExtent.yMinimum() ) / outputCellSizeY );
+
+  const QString outputFile = parameterAsOutputLayer( parameters, QStringLiteral( "OUTPUT" ), context );
+  const QFileInfo fi( outputFile );
+  const QString outputFormat = QgsRasterFileWriter::driverForExtension( fi.suffix() );
+
+  auto writer = std::make_unique<QgsRasterFileWriter>( outputFile );
+  writer->setOutputFormat( outputFormat );
+  std::unique_ptr<QgsRasterDataProvider> provider( writer->createMultiBandRaster( outputDataType, cols, rows, outputExtent, outputCrs, mRanks.size() ) );
+  if ( !provider )
+    throw QgsProcessingException( QObject::tr( "Could not create raster output: %1" ).arg( outputFile ) );
+  if ( !provider->isValid() )
+    throw QgsProcessingException( QObject::tr( "Could not create raster output %1: %2" ).arg( outputFile, provider->error().message( QgsErrorMessage::Text ) ) );
+  provider->setNoDataValue( 1, outputNoData );
+
+  std::map<QString, std::unique_ptr<QgsRasterInterface>> inputInterfaces;
```

Comment:
```suggestion
  std::map<QString, std::unique_ptr<QgsRasterInterface>> newProjectorInterfaces;
  std::map<QString, QgsRasterInterface*> inputInterfaces;
```

Reply:
Hm, I'm not sure it's a good rename here. That map will hold either new projector or preexisting data provider, I feel renaming it to newProjectorInterfaces fails to reflect the latter. Thoughts?

Reply:
It won't have pre-existing providers -- those are only references in inputInterfaces, whereas newProjectorInterfaces is the container for newly created and owned objects only.

Reply:
Oh I see what you are suggesting now, the way github highlighted your suggestion misled me.